### ISSUE TITLE
MD031 fails on nested (example) code blocks

### ISSUE THREAD

**Issue by waylan** [markdownlint/markdownlint#100]
Sometimes a code block may contain an example of a code block. For example when documenting the Markdown syntax, like this:

``````
~~~no-highlight
```
Fenced code blocks are like Standard
Markdown’s regular code blocks, except that
they’re not indented and instead rely on
start and end fence lines to delimit the
code block.
```
~~~
``````

The above generates the following errors:

``` bash
$ mdl foo.txt 
foo.txt:2: MD031 Fenced code blocks should be surrounded by blank lines
foo.txt:8: MD031 Fenced code blocks should be surrounded by blank lines
```

I've tried changing the style of the chars (backticks or tildes), changing the number used (3 wrapping 4 and 4 wrapping 3), changing the language def (space before it or none) and is seems to make no difference. Of course, the file is rendered to HTML just fine.

Finally, I also tried adding a blank line before and after the nested code block, like this:

``````
~~~no-highlight

```
Fenced code blocks are like Standard
Markdown’s regular code blocks, except that
they’re not indented and instead rely on
start and end fence lines to delimit the
code block.
```

~~~
``````

And that made no difference: 

``` bash
$ mdl foo.txt 
foo.txt:3: MD031 Fenced code blocks should be surrounded by blank lines
foo.txt:9: MD031 Fenced code blocks should be surrounded by blank lines
```

However, it did add an unwanted blank line to the start and end of the code block in the rendered HTML, so that is **not** the correct solution.

I realize I could indent the fenced block to nest it in an indented block, however, with the JavaScript based highlighter being used, I need to tell it not to highlight this block (or it will try to guess incorrectly and muck it all up). Therefore, I need a fence code block nested in a fenced code block.

I could disable rule MD031, but this is actually a desirable rule. It is simply failing to account for this edge  case, IMO. Therefore, I'd like to use this rule, but can't as long as it fails for this edge case.

**Comment by waylan** [comment]
I would like to add that the Kradown documentation specifically [mentions](http://kramdown.gettalong.org/syntax.html#fenced-code-blocks) that Kramdown can parse nested fenced code blocks. Therefore, I am assuming that this is solely a bug in the linter which is not accounting for this (very common in my experience) use case.

**Comment by DavidAnson** [comment]
Rule MD031 doesn't use Kramdown because it can interfere with what the rule tries to detect (I originally authored this rule, so the bug is my fault). The line-by-line logic the code uses instead doesn't correctly match the close of the outer fence, thinking instead that there are two separate empty fenced sections at the top/bottom. I don't think it would be all that hard to fix the rule by matching the fence start marker explicitly as required by convention, but I haven't had time for projects lately. If you need to work around this for now, I think the below will work (though it's not quite equivalent):

``````
~~~no-highlight
```

Fenced code blocks are like Standard
Markdown’s regular code blocks, except that
they’re not indented and instead rely on
start and end fence lines to delimit the
code block.

```
~~~
``````

My other thought would be to use indenting instead of a fence for the outer code block (~ in your example), though that means you'd lose the ability to specify the highlight language:

``````
    ```
    Fenced code blocks are like Standard
    Markdown’s regular code blocks, except that
    they’re not indented and instead rely on
    start and end fence lines to delimit the
    code block.
    ```
``````

Sorry for the trouble - I hope this helps!

PS - In typing the above, it seems that maybe GitHub's Markdown parser has issues with nesting - though I'm in no position to cast stones. :)

**Comment by DavidAnson** [comment]
FYI, I have just added code for handling nested fences to the Node.js implementation of markdownlint: DavidAnson/markdownlint@bb0d8a36b320e2f295388ee3cc0a0f638f12c7eb. The new test case is directly usable by the Ruby markdownlint test suite; the change itself is relatively simple. It handles nesting per the CommonMark spec (link in the commit message), though it does not add support for indentation of fences. If nobody beats me to it, I'll look at creating a pull request for this project in the near future.

**Comment by waylan** [comment]
Awesome. Thanks for fixing this. Any idea when this would be available in a release? Not trying to be pushy. Just curious.

**Comment by mivok** [comment]
Hi Waylan,

There is no formal release schedule currently (right now I'm just one person working on this in my spare time). That said, if there are no major issues, I plan to get a release out in the next few days.
